src/libertem/udf/base.py

```diff
@@ -456,6 +471,36 @@ def set_meta(self, meta):
     def set_slice(self, slice_):
         self.meta.slice = slice_
 
+    def set_coordinates(self, slice_, ds_shape, roi):
+        o = slice_.origin
+        s = slice_.shape
+        sig_dims = s.sig.dims
+        start_idx = o[0]
+        end_idx = o[0] + s[0]
+        nav_shape = ds_shape[:-sig_dims]
+        if roi is None:
+            flat_nav_shape = tuple((int(np.prod(nav_shape)),))
+            self.meta.coordinates = [
+                np.unravel_index(
+                    np.ravel_multi_index(tuple((i,)), flat_nav_shape),
+                    nav_shape
+                ) for i in range(start_idx, end_idx)
+            ]
+        else:
+            self.meta.coordinates = []
+            ds_shape = Shape(ds_shape, sig_dims=sig_dims)
+            ds_slice = Slice(origin=[0] * len(ds_shape), shape=ds_shape)
+            # in case roi is flattened
+            if roi.shape != nav_shape:
+                roi = roi.reshape(nav_shape)
+            indices = _roi_to_nd_indices(roi, ds_slice)
+            for idx, frame_idx in enumerate(indices):
+                if idx in range(start_idx, end_idx):
+                    self.meta.coordinates.append(frame_idx)
+                # no more frames from the ROI in the slice
+                elif idx >= end_idx:
+                    break
```

Comment:
Hmm, we should see if this is possible without having this loop here, as it will always iterate over all indices for `ds_slice`. Isn't it possible to just slice `indices`?